CTFE interning: don't walk allocations that don't need it

lqd opened the pull request:
The interning of const allocations visits the mplace looking for references to intern. Walking big aggregates like big static arrays can be costly, so we only do it if the allocation we're interning contains references or interior mutability.

Walking ZSTs was avoided before, and this optimization is now applied to cases where there are no references/relocations either.

---

While initially looking at this in the context of #93215, I've been testing with smaller allocations than the 16GB one in that issue, and with different init/uninit patterns (esp. via padding).

In that example, by default, `eval_to_allocation_raw` is the heaviest query followed by `incr_comp_serialize_result_cache`. So I'll show numbers when incremental compilation is disabled, to focus on the const allocations themselves at 95% of the compilation time, at bigger array sizes on these minimal examples like `static ARRAY: [u64; LEN] = [0; LEN];`. 

That is a close construction to parts of the `ctfe-stress-test-5` benchmark, which has const allocations in the megabytes, while most crates usually have way smaller ones. This PR will have the most impact in these situations, as the walk during the interning starts to dominate the runtime. 

Unicode crates (some of which are present in our benchmarks) like `ucd`, `encoding_rs`, etc come to mind as having bigger than usual allocations as well, because of big tables of code points (in the hundreds of KB, so still an order of magnitude or 2 less than the stress test).

In a check build, for a single static array shown above, from 100 to 10^9 u64s (for lengths in powers of ten), the constant factors are lowered:

(log scales for easier comparisons)
![plot_log](https://user-images.githubusercontent.com/247183/171422958-16f1ea19-3ed4-4643-812c-1c7c60a97e19.png)

(linear scale for absolute diff at higher Ns)
![plot_linear](https://user-images.githubusercontent.com/247183/171401886-2a869a4d-5cd5-47d3-9a5f-8ce34b7a6917.png)

For one of the alternatives of that issue
```rust
const ROWS: usize = 100_000;
const COLS: usize = 10_000;

static TWODARRAY: [[u128; COLS]; ROWS] = [[0; COLS]; ROWS];
```

we can see a similar reduction of around 3x (from 38s to 12s or so).

For the same size, the slowest case IIRC is when there are uninitialized bytes e.g. via padding

```rust
const ROWS: usize = 100_000;
const COLS: usize = 10_000;

static TWODARRAY: [[(u64, u8); COLS]; ROWS] = [[(0, 0); COLS]; ROWS];
```
then interning/walking does not dominate anymore (but means there is likely still some interesting work left to do here). 

Compile times in this case rise up quite a bit, and avoiding interning walks has less impact: around 23%, from 730s on master to 568s with this PR.

rust-highfive commented:
Some changes occured to the CTFE / Miri engine

cc @rust-lang/miri

oli-obk commented on compiler/rustc_const_eval/src/interpret/memory.rs:
```diff
@@ -935,6 +935,10 @@ impl<'tcx, 'a, Tag: Provenance, Extra> AllocRef<'a, 'tcx, Tag, Extra> {
             .map_err(|e| e.to_interp_error(self.alloc_id))?)
     }
+
+    pub(crate) fn get_relocations(&self, range: AllocRange) -> &[(Size, Tag)] {
+        self.alloc.get_relocations(&self.tcx, self.range.subrange(range))
```
You can remove the `range` argument and just always use `self.range` internally. It's the only way this is used anyway.

lqd replied:
I've amended the commit to do this change.

Thanks oli. Yosh *[is right](https://twitter.com/yoshuawuyts/status/1531654726512693250)*.

lqd commented:
Opening as draft as I still need to ensure test coverage is sufficient, and add some if necessary. It worked locally in my small scale tests.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 6c492abbbc94e85c545e996922b0275b4580265f with merge 0aa3f5e10899462a2f253e01eb637fa40eee7095...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6c492abbbc94e85c545e996922b0275b4580265f","merge_sha":"0aa3f5e10899462a2f253e01eb637fa40eee7095"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6675374370?check_suite_focus=true)
Build commit: 0aa3f5e10899462a2f253e01eb637fa40eee7095 (`0aa3f5e10899462a2f253e01eb637fa40eee7095`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6675374370?check_suite_focus=true"},"merge_sha":"0aa3f5e10899462a2f253e01eb637fa40eee7095"} -->

rust-timer commented:
Queued 0aa3f5e10899462a2f253e01eb637fa40eee7095 with parent 16a0d03698bfc9f93250490797f9a1a870f8bcfe, future [comparison URL](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095).

rust-timer commented:
Finished benchmarking commit (0aa3f5e10899462a2f253e01eb637fa40eee7095): [comparison url](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095&stat=instructions:u)

- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -2.8%    | -9.9% | 26        |
| All 😿🎉 (primary) | N/A      | N/A | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 🎉 relevant improvement found
- Secondary benchmarks: 😿 relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 3.7%     | 3.7% | 1         |
| Improvements 🎉 <br /> (primary) | -2.3%    | -2.3% | 1         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -2.3%    | -2.3% | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.7%     | 2.7% | 1         |
| Regressions 😿 <br /> (secondary) | 2.4%     | 2.7% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -5.4%    | -6.1% | 4         |
| All 😿🎉 (primary) | 2.7%     | 2.7% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

lqd commented on compiler/rustc_const_eval/src/interpret/memory.rs:
```diff
@@ -935,6 +935,10 @@ impl<'tcx, 'a, Tag: Provenance, Extra> AllocRef<'a, 'tcx, Tag, Extra> {
     }
+
+    pub(crate) fn get_relocations(&self, range: AllocRange) -> &[(Size, Tag)] {
+        self.alloc.get_relocations(&self.tcx, self.range.subrange(range))
```
(Probably a dupe from the comment above, so I'll resolve both at the same time)

oli-obk commented:
cc @RalfJung this PR does a work-skipping in the interner that is mostly copied from validation

oli-obk commented:
r? @RalfJung as I worked with lqd directly on the impl

RalfJung commented:
> While initially looking at this in the context of https://github.com/rust-lang/rust/issues/93215, I've been testing with smaller allocations than the 16GB one in that issue, and with .

That sentence in the PR description just ends mid-way?

RalfJung commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,10 +168,22 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
+        // also need to take special care of interior mutability.
+        //
+        // As an optimization, however, if the allocation does not contain any references: we don't
```
```suggestion
        // As an optimization, however, if the allocation does not contain any pointers: we don't
```

lqd replied:
Thank you, I've incorporated that change.

RalfJung commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,10 +168,22 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+        let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
+            // We could be dealing with an extern type here in the future, so we do the regular
+            // walk.
+            return self.walk_aggregate(mplace, fields);
```
We are skipping the `UnsafeCell` logic here. Please add a comment explaining why that is okay.

lqd replied:
I've done this when addressing the next comment's points, with the refactor you suggested.

lqd commented on compiler/rustc_const_eval/src/interpret/memory.rs:
```diff
@@ -935,6 +935,11 @@ impl<'tcx, 'a, Tag: Provenance, Extra> AllocRef<'a, 'tcx, Tag, Extra> {
     }
+
+    /// Returns the allocation's relocations for the entire range of the `AllocRef`.
+    pub(crate) fn get_relocations(&self) -> &[(Size, Tag)] {
```
Agreed. I’ll change it to "has_relocations".

lqd replied:
I've done that change to limit the API surface to having relocations.

And I've also tried to fix the possible oversight you mention in https://github.com/rust-lang/rust/pull/97585/commits/c792460985e14fb5d674523e5d909a39692dc8db, by making `get_relocations` private. 

I don't know if that's what you had in mind to hide the relocations behind the `Allocation` abstraction ? (as I'm not super aware of the difference between accessing relocations using `get_relocations` compared to the public [`relocations` getter](https://github.com/rust-lang/rust/blob/083721a1a7365d3afe1521cd2661b2201aac0450/compiler/rustc_middle/src/mir/interpret/allocation.rs#L259-L262))

lqd commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -186,6 +198,11 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+        if alloc.get_relocations().is_empty() {
+            // There are no refs or relocations in this allocation, we can skip the interning walk.
+            return Ok(());
+        }
```
To not change the existing process too much (keeping the existing order of checks: ZST, then UnsafeCell) and avoid the optimization if any interior mutability happens in general: we thought the complexity of handling both would not be worth it just for this rare combination. 

Is that description accurate @oli-obk ?

RalfJung replied:
Well, but you did add some new short-circuiting that happens before the UnsafeCell path. And The UnsafeCell path is also only relevant for the references we encounter in there.

In the end, what all your logic here should do is compute a single `bool`: "can we skip this?". I suggest to structure the code accordingly (e.g. in a helper closure, so that you can easily short-circuit that computation). That `bool` should be `false` if we cannot determine the size. But if we decide that we can indeed skip, then we should do that before even testing for UnsafeCell.

lqd replied:
Are you referring to the extern types early return as "new short circuiting" ? I’m not sure this has changed in practice, the UnsafeCell ty would not match in this situation, right ?

RalfJung replied:
I think it might match, in situations like `UnsafeCell<(&T, ExternType)>`.

lqd replied:
To err on the safer side and to cleanup the function, I've done the refactor you suggest, in https://github.com/rust-lang/rust/pull/97585/commits/4bcb6118ff6fbe091c1cc5d7ea306bdc9e39796c

RalfJung commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,10 +168,22 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+        // As an optimization, however, if the allocation does not contain any references: we don't
+        // need to do the walk. It can be costly for big arrays for example (e.g. issue #93215).
+
+        let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
```
FWIW, this call here might be not entirely cheap. So doing it recursively on all aggregates we encounter might be a bad idea. Maybe leave a FIXME?

An alternative might be to only go check this if the type is `Array` or `Slice`, as those are usually the large types.

lqd replied:
I've implemented this narrowing in https://github.com/rust-lang/rust/pull/97585/commits/84bc03f1a1316abd1ae3ee0b345c93b11c6cd579.

lqd commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -179,21 +179,25 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
-                return Ok(false);
+                // If there are no refs or relocations in this allocation, we can avoid the
+                // interning walk.
+                if let Some(alloc) = self.ecx.get_ptr_alloc(mplace.ptr, size, align)? {
+                    if !alloc.has_relocations() {
+                        return Ok(false);
+                    }
+                }
```
A couple things here:
- I'm not using a let chain because rustfmt ignores it, and doesn't format this block if I do.
- if we fail to get an `AllocRef`, we will do the walk. I believe we could equally not do it, since there shouldn't be any relocations anyways, but it shouldn't matter in practice IIUC ?

lqd commented:
> That sentence in the PR description just ends mid-way?

I've fixed that.

@RalfJung Thanks for the review suggestions, I've pushed commits to address them (limiting the `has_relocations` API surface I've added, as well as the existing `get_relocations` public API, turning the function into a single early return point to skip the walk, etc)

Does that look more like what you had in mind ?

@rustbot ready

Let's see if the `ctfe-stress-5` benchmark still looks like the previous run did:
@bors try @rust-timer queue

lqd commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+
+                // If there are no refs or relocations in this allocation, we can avoid the
+                // interning walk.
+                if let Some(alloc) = self.ecx.get_ptr_alloc(mplace.ptr, size, align)? {
+                    if !alloc.has_relocations() {
+                        return Ok(false);
+                    }
+                }
```
A couple things here:

- I'm not using a let chain because rustfmt ignores it, and doesn't format this block if I do.
- if we fail to get an AllocRef, we will do the walk. I believe we could equally not do it, since there shouldn't be any relocations anyways, but it shouldn't matter in practice IIUC ?

RalfJung replied:
Yeah if `get_ptr_alloc` returns `None` then this is a ZST. So better to not recurse, that seems less confusing.

lqd replied:
I'll do that, then, thank you. 

Interestingly I think I saw some `None`s here that didn't apparently hit the ZST early return at the beginning of the function, would you find this unusual or completely fine ?

RalfJung replied:
That sounds odd. Basically it should be the case that if `get_ptr_alloc` returns `None` then `size` must be 0.

However I guess `size` being 0 does not mean it is a ZST... it could be an empty slice. Dynamically sized types are not considered ZST. So yeah I guess this is fine.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6866615246?check_suite_focus=true)
Build commit: 22b3b60b85aaa0354c4f4e49024750163946f44d (`22b3b60b85aaa0354c4f4e49024750163946f44d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6866615246?check_suite_focus=true"},"merge_sha":"22b3b60b85aaa0354c4f4e49024750163946f44d"} -->

rust-timer commented:
Queued 22b3b60b85aaa0354c4f4e49024750163946f44d with parent 083721a1a7365d3afe1521cd2661b2201aac0450, future [comparison URL](https://perf.rust-lang.org/compare.html?start=083721a1a7365d3afe1521cd2661b2201aac0450&end=22b3b60b85aaa0354c4f4e49024750163946f44d).

rust-timer commented:
Finished benchmarking commit (22b3b60b85aaa0354c4f4e49024750163946f44d): [comparison url](https://perf.rust-lang.org/compare.html?start=083721a1a7365d3afe1521cd2661b2201aac0450&end=22b3b60b85aaa0354c4f4e49024750163946f44d&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=083721a1a7365d3afe1521cd2661b2201aac0450&end=22b3b60b85aaa0354c4f4e49024750163946f44d&stat=instructions:u)

- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 0.5%     | 0.6% | 4         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.4%    | -10.3% | 21        |
| All 😿🎉 (primary) | 0.5%     | 0.6% | 4         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=083721a1a7365d3afe1521cd2661b2201aac0450&end=22b3b60b85aaa0354c4f4e49024750163946f44d&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: 🎉 relevant improvement found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.2%     | 2.2% | 1         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -4.0%    | -4.0% | 1         |
| All 😿🎉 (primary) | 2.2%     | 2.2% | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=083721a1a7365d3afe1521cd2661b2201aac0450&end=22b3b60b85aaa0354c4f4e49024750163946f44d&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -6.0%    | -7.5% | 5         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

lqd commented:
Unfortunately, the command to run the benchmark under cg diff shown on the compare page doesn't work locally.

```
==82821== Valgrind: debuginfo reader: Possibly corrupted debuginfo file.
==82821== Valgrind: I can't recover.  Giving up.  Sorry.
```

(are you OK, valgrind ?)

Checking with a local build, so not exactly the same thing as perf.rlo runs, it seems the changes in `diesel` are mostly related to memory allocations. A bit unexpected as I don't believe this changed since the previous run. 

Unfortunately, the previous run's commit also has the same profile locally as the new run, so it's hard to say what the perfbot sees as different here, especially since it's not always possible to run cg locally on the actual CI artifacts. (Diesel could also be a bit noisy recently ?)

RalfJung commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+                return Ok(false);
+            }
+
+            // Now, check whether this allocation contains reference types (as relocations).
```
```suggestion
            // Now, check whether this allocation contains pointers (as relocations).
```
This is not about the types, it's about the actual values.

lqd replied:
I will change this but I was copying from the few lines below which contain both "references" and "reference types" https://github.com/rust-lang/rust/blob/94e93749ab00539a11e90426ea87382c433530a8/compiler/rustc_const_eval/src/interpret/intern.rs#L176-L194 should I change something else in the file ?

RalfJung replied:
It does not contain "reference types".

But yes it does say "references" and... indeed the internet only looks for references, not raw pointers. So let's agree on "references"?

lqd replied:
It does at the bottom, here it as as a single link so that it's not hidden by the scrollbar: https://github.com/rust-lang/rust/blob/94e93749ab00539a11e90426ea87382c433530a8/compiler/rustc_const_eval/src/interpret/intern.rs#L192-L195

RalfJung replied:
Ah yes, but that code is talking about `ty::Ref` -- reference types.

lqd replied:
I can change all the "pointers" to "references" if you'd prefer (or the opposite) ? I still need to launch another perf run after this one anyways, to see if the original commit shows change on `diesel` and could do that at the same time.

RalfJung replied:
The thing is, "whether this allocation contains reference types (as relocations)" is meaningless. Relocations represent reference (and pointer) *values*. And it is the (reference) *values* that we are after here, the types just guide us there. But there being no relocations does *not* mean there there are no reference types! It just means that there are no reference *values*.

RalfJung replied:
Ah, I only just saw you already changed that  -- sorry.

So, yeah, it should probably say references in some of the places it currently says pointers. But at least some of the existing code is right when it talks about pointers... so if you change it, only change it in comments you added.

lqd replied:
Will do. 

I'll also add clarifications to the other comments I added elsewhere, to match your latest comment which was quite clear. Thanks for taking the time to explain.

lqd replied:
To ensure I understood correctly (it's late here), @RalfJung, IIUC the following would be acceptable, right ?

> Now, check whether this allocation could contain references.
>
> If there are no relocations in this allocation, it does not contain references and we can avoid the interning walk.

RalfJung replied:
Seems fine... if I am super nitpicky we'd have to go into detail about pointer and integer values and how values of reference types can be either, but maybe let's not. ;)

Maybe "does not contain references that point to another allocation"? That's hopefully still understandable and also technically correct. ;)

(The concern is references like `&*(1usize as *const ())`, which is a perfectly fine value of type `&()` but does not involve a relocation.)

lqd replied:
I've pushed a commit to address this.

RalfJung commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
     ) -> InterpResult<'tcx> {
-        // ZSTs cannot contain pointers, so we can skip them.
-        if mplace.layout.is_zst() {
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
```
```suggestion
        // We want to walk the aggregate to look for pointers to intern. While doing that we
```

lqd replied:
I've pushed a commit to address this.

lqd commented on compiler/rustc_const_eval/src/interpret/intern.rs:
```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
+            // to avoid could be expensive: on the potentially larger types, arrays and slices,
+            // rather than on all aggregates unconditionally.
+            if matches!(mplace.layout.ty.kind(), ty::Array(..) | ty::Slice(..)) {
+                let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
```
valgrind have finally fixed the dwarf bug and I was able to run the CI artifacts to reproduce the issue it sees, that oli posted.

> The only possible source of the regression that I can see

I wonder if that's the case. `diesel` is unusual here in that 3% of the check time is spent in `try_fold_with` and the likes, does the current PR change things related to folding at all @oli-obk ? (I would expect it not to).

---

I'm seeing differences locally as well, but they are completely unrelated (and it seems unlikely it's about PGO) but could be about allocation: cachegrind attributes more time to ` ./string/../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:__memcpy_avx_unaligned_erms`, the problem is that it does this here even for two different rustc worktrees for _the same commit and same config.toml_, making things though to analyze (I'll ignore this memcpy difference).

```
```

The `visit_aggregate` function in particular is pretty short, and some variance in the machine makes the 4945 calls' wall-time noisy (and doesn't register in cachegrind at all, even when preventing inlining) like so:
- master: 19ms
- this PR: 16ms

The aggregates look like this:
- total:
- ZSTs early return:
-

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 2344e664aae22c0a3c29e4caa26fa51db4e1a6a4 with merge f983b52da2d21c5393097bc352068af1ebe07bab...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2344e664aae22c0a3c29e4caa26fa51db4e1a6a4","merge_sha":"f983b52da2d21c5393097bc352068af1ebe07bab"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7102156017?check_suite_focus=true)
Build commit: f983b52da2d21c5393097bc352068af1ebe07bab (`f983b52da2d21c5393097bc352068af1ebe07bab`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7102156017?check_suite_focus=true"},"merge_sha":"f983b52da2d21c5393097bc352068af1ebe07bab"} -->

rust-timer commented:
Queued f983b52da2d21c5393097bc352068af1ebe07bab with parent 94e93749ab00539a11e90426ea87382c433530a8, future [comparison URL](https://perf.rust-lang.org/compare.html?start=94e93749ab00539a11e90426ea87382c433530a8&end=f983b52da2d21c5393097bc352068af1ebe07bab).

rust-timer commented:
Finished benchmarking commit (f983b52da2d21c5393097bc352068af1ebe07bab): [comparison url](https://perf.rust-lang.org/compare.html?start=94e93749ab00539a11e90426ea87382c433530a8&end=f983b52da2d21c5393097bc352068af1ebe07bab&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=94e93749ab00539a11e90426ea87382c433530a8&end=f983b52da2d21c5393097bc352068af1ebe07bab&stat=instructions:u)

- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 0.4%     | 0.5% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.0%    | -8.9% | 21        |
| All 😿🎉 (primary) | N/A      | N/A | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=94e93749ab00539a11e90426ea87382c433530a8&end=f983b52da2d21c5393097bc352068af1ebe07bab&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 2.5%     | 2.8% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.5%    | -3.6% | 2         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=94e93749ab00539a11e90426ea87382c433530a8&end=f983b52da2d21c5393097bc352068af1ebe07bab&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 2.7%     | 2.7% | 1         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -6.7%    | -7.9% | 7         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

lqd commented:
Alright, the `diesel` regression seems like it was noise indeed. I won't need to do other perf runs to tackle it.

Apart from possibly some more wording changes you'd want to see, this PR looks ready to me.

RalfJung commented:
There's still a .5% regression in "[deeply-nested-multi](https://github.com/rust-lang/rustc-perf/tree/master/collector/benchmarks/deeply-nested-multi)", whatever that is about. But the speedups are very nice. :)[deeply-nested-multi](https://github.com/rust-lang/rustc-perf/tree/master/collector/benchmarks/deeply-nested-multi)

RalfJung commented:
Comments look good to me. :)

lqd commented:
> There's still a .5% regression in "[deeply-nested-multi](https://github.com/rust-lang/rustc-perf/tree/master/collector/benchmarks/deeply-nested-multi)", whatever that is about. But the speedups are very nice. :)

Yeah I saw that. That benchmark looks a bit noisy as of late:
![image](https://user-images.githubusercontent.com/247183/176325723-c4dc6196-223b-40bf-a9f6-0a4313c4a8b7.png)

The regression wasn't present in the previous runs of this PR, and when that benchmark did appear, it was [a -1% win](https://perf.rust-lang.org/compare.html?start=16a0d03698bfc9f93250490797f9a1a870f8bcfe&end=0aa3f5e10899462a2f253e01eb637fa40eee7095&stat=instructions:u), likely noise again.

Coupled to the fact that this is a stress test "secondary benchmark", the perfbot does not even count this as a noteworthy regression, so I won't either 🤡.

lqd commented:
@RalfJung is there anything more you'd like me to do in this PR ? Or is it good enough for a r+ ? :)

RalfJung commented:
Sorry, I didn't realize I am the assigned reviewer.^^
Looks like we are happy with the perf, so
@bors r+

bors commented:
:pushpin: Commit d634f14f26dc6640a9b6cd8b5f896aaf1ecc3cb6 has been approved by `RalfJung`

<!-- @bors r=RalfJung d634f14f26dc6640a9b6cd8b5f896aaf1ecc3cb6 -->
<!-- homu: {"type":"Approved","sha":"d634f14f26dc6640a9b6cd8b5f896aaf1ecc3cb6","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit d634f14f26dc6640a9b6cd8b5f896aaf1ecc3cb6 with merge 750d6f85459356db4838dc06db8b19406e1ed31a...
<!-- homu: {"type":"BuildStarted","head_sha":"d634f14f26dc6640a9b6cd8b5f896aaf1ecc3cb6","merge_sha":"750d6f85459356db4838dc06db8b19406e1ed31a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7164084511?check_suite_focus=true)
Approved by: RalfJung
Pushing 750d6f85459356db4838dc06db8b19406e1ed31a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7164084511?check_suite_focus=true"},"merge_sha":"750d6f85459356db4838dc06db8b19406e1ed31a"} -->

rust-timer commented:
Finished benchmarking commit (750d6f85459356db4838dc06db8b19406e1ed31a): [comparison url](https://perf.rust-lang.org/compare.html?start=6a1092056441652fe5fe5c5b422644951e6b99ce&end=750d6f85459356db4838dc06db8b19406e1ed31a&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6a1092056441652fe5fe5c5b422644951e6b99ce&end=750d6f85459356db4838dc06db8b19406e1ed31a&stat=instructions:u)

- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -1.0%    | -1.0% | 2         |
| Improvements 🎉 <br /> (secondary) | -2.5%    | -9.3% | 32        |
| All 😿🎉 (primary) | -1.0%    | -1.0% | 2         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6a1092056441652fe5fe5c5b422644951e6b99ce&end=750d6f85459356db4838dc06db8b19406e1ed31a&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 1.1%     | 1.1% | 1         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -2.8%    | -2.8% | 1         |
| Improvements 🎉 <br /> (secondary) | -2.0%    | -2.4% | 2         |
| All 😿🎉 (primary) | -0.9%    | -2.8% | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6a1092056441652fe5fe5c5b422644951e6b99ce&end=750d6f85459356db4838dc06db8b19406e1ed31a&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 3.0%     | 4.1% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -6.1%    | -7.6% | 7         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
